Add `race-network-and-fetch-handler` option

sisidovski opened the pull request:
This PR will add the spec explanation for the `race-network-and-fetch-handler` source in the static routing API.

This option starts the race between the network request and the fetch handler execution, and use the result whichever comes first. Currently the result from the network request is used only when the the response is successfully returned. 

Borrowed sentences mostly from the navigation preload.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/sisidovski/ServiceWorkerForStaticRoutingAPI/pull/10.html" title="Last updated on Feb 16, 2024, 3:59 AM UTC (bf4bdaf)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/10/59e1de9...sisidovski:bf4bdaf.html" title="Last updated on Feb 16, 2024, 3:59 AM UTC (bf4bdaf)">Diff</a>

sisidovski committed:
Add `race-network-and-fetch-handler` option

sisidovski commented:
@yoshisatoyanagisawa @domenic Could you take a look? As this is the first draft, I'm not confident with how to describe the parallelism on this option, and want to know what's missing.

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -3145,6 +3145,23 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+              Note: In order to avoid duplicated fetches to |request|, the user agent may reuse |raceNetworkRequestResponse| as the result of |e|'s [=FetchEvent/potential response=], or the fallback request.
+
+              1. Let |raceFetchController| be null.
```
Don't we need to abort |raceFetchController| when "[Dispatch](https://dom.spec.whatwg.org/#concept-event-dispatch) e at activeWorker’s [global object](https://w3c.github.io/ServiceWorker/#dfn-service-worker-global-object)." succeeds?

sisidovski replied:
Good catch. 

If the fetch handler result is fallback, we may want to reuse the inflight request as the fallback request.
If the fetch handler responded, the race fetch is not needed anymore, so we can abort it.

yoshisatoyanagisawa replied:
I just wondered the case for race condition.
I suppose |response| can be overwritten by raceFetchController's response if raceFetchController's abort happened after getting response from it.

Even if |raceFetchController| side wins the competition, fetch event side seems not be aborted.  Is it fine to return |response| twice?

sisidovski replied:
I believe we should complete the fetch handler execution rather than abort it. But you're right, `Handle Fetch` shouldn't return the response twice.

Updated `Handle Fetch` to manage the race result with queue, which return |response| only once.

yoshisatoyanagisawa replied:
Thanks for the update.  I think the race condition is resolved.

sisidovski committed:
Abort the request if the fetch handler respond

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -3145,6 +3146,22 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
                   1. Return |response|.
+          1. Else if |source| is {{RouterSource/"race-network-and-fetch-handler"}}, and |request|'s [=request/method=] is \`<code>GET</code>\` then:
+
+              Note: In order to avoid duplicated fetches to |request|, the user agent may reuse |raceNetworkRequestResponse| as the result of |e|'s [=FetchEvent/potential response=], or the fallback request.
```
nit but |e| might not be defined yet.  I guess it fair to say `FetchEvent` here.

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -3145,6 +3146,22 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                          1. If |raceNetworkRequestResponse|'s [=response/status=] is `200`, then:
+                            1. Set |response| to |raceNetworkRequestResponse|.
+                            1. Return |response|.
+          1. [=If aborted=], then:
```
Maybe wrong indent?  You might need another indent here, or this statement is the same level with dispatching behaviors for source.

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -3145,6 +3146,22 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              Note: In order to avoid duplicated fetches to |request|, the user agent may reuse |raceNetworkRequestResponse| as the result of |e|'s [=FetchEvent/potential response=], or the fallback request.
+
+              1. Run the following substeps [=in parallel=], but [=abort when=] |controller|'s [=fetch controller/state=] is "<code>terminated</code>" or "<code>aborted</code>":
+                  1. Set |raceFetchController| to the result of [=Fetch|fetching=] |request|.
```
How |raceFetchController|'s abort is handled?

sisidovski replied:
It's just aborted, but not handled. The network request here is an additional request. `Create Fetch Event and Dispatch` has the step that informs the abort signal to the controller, which is `[=Queue a task=] to [=AbortController/signal abort=] on |controller| with |deserializedError|.`

I don't have a clear thoughts if we should send a signal to the controller like that, or it's fine to do nothing. Actually this is the same as what navigation preload does below.

domenic replied:
I *think* how this works is that aborting a fetch controller has two impacts:

1. Terminates the network connection. Visible to the server.
2. Communicates to any `fetch()` callers the appropriate promise rejection.

(2) is not applicable here since we are performing the fetch from the spec directly, instead of from the `fetch()` method.

So I think this is OK as written. Aborting is good because it can save server resources, but doesn't have many observable affects, and for example the error that you use doesn't matter.

sisidovski replied:
Thanks, I'll keep this as it is.

sisidovski committed:
Fix nav preload condition, indent, FetchEvent

sisidovski committed:
Split Handle Fetch

sisidovski committed:
Fix id name and tags

sisidovski committed:
Restore old sentence

sisidovski commented:
Updated to address review points and some concerns.
Main diff:
- Fixed not to return |response| twice on `Handle Fetch`, by splitting `Handle Fetch` and call the latter part with `in parallel`, and manage the race with queue.
- If `race-network-and-fetch-handler` source is used, do not dispatch navigation preload.

Note: In the race, the response from the network request is queued only when the result is successful (currently 200). On the other hand, the fetch handler response is queued regardless of the result.

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -3154,7 +3150,30 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                          1. [=queue/Enqueue=] |raceNetworkRequestResponse| to |queue|.
+              1. [=If aborted=], then:
+                  1. Let |deserializedError| be the result of [=deserialize a serialized abort reason=] given null and |workerRealm|.
+                  1. [=fetch controller/Abort=] |raceFetchController| with |deserializedError|.
```
Maybe only if |raceFetchController| is not null?

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -3154,7 +3150,30 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                  1. [=queue/Enqueue=] |fetchHandlerResponse| to |queue|.
+              1. Wait until |queue| is not empty.
+              1. Set |response| to the result of [=dequeue=] |queue|.
+              1. Return |response|.
```
Just before this step, don't we abort |raceFetchController| if it is still running?

sisidovski replied:
We may want to reuse the response from the network request as a fallback. Updated to abort it only when |fetchHandlerResponse| is a successful response. This aligns with the current chromium implementation.
https://source.chromium.org/chromium/chromium/src/+/main:content/browser/service_worker/service_worker_main_resource_loader.cc;l=722;drc=7dc3003765c74daa4bb49e4463e3a7963d8c7ea1;bpv=1;bpt=1?q=main_resource_loader&ss=chromium%2Fchromium%2Fsrc

yoshisatoyanagisawa commented:
Other than my minor comments, the PR looks good to me.

sisidovski committed:
Abort race request if not needed

sisidovski commented:
@domenic Could you take a look? Thanks!

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,33 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. Let |raceFetchController| be null.
+              1. Run the following substeps [=in parallel=], but [=abort when=] |controller|'s [=fetch controller/state=] is "<code>terminated</code>" or "<code>aborted</code>":
+                  1. Set |raceFetchController| to the result of [=Fetch|fetching=] |request|.
+                  1. To [=fetch/processResponse=] for |raceNetworkRequestResponse|, run these substeps:
```
Write this using the modern style shown in https://fetch.spec.whatwg.org/#example-callback-chunk-by-chunk

sisidovski replied:
Updated. Does this look good?

domenic replied:
Looks good!

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,33 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                      1. If |raceNetworkRequestResponse|'s [=response/type=] is "`error`", then:
+                          1. Terminate these substeps.
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is `200`, then:
+                          1. [=queue/Enqueue=] |raceNetworkRequestResponse| to |queue|.
```
These four steps seem overly complicated to me. I would suggest only one step:

1. If _raceNetworkRequestResponse_'s status is 200, then enqueue _raceNetworkRequestResponse_ to _queue_.

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,33 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                  1. To [=fetch/processResponse=] for |raceNetworkRequestResponse|, run these substeps:
+                      1. If |raceNetworkRequestResponse|'s [=response/type=] is "`error`", then:
+                          1. Terminate these substeps.
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is `200`, then:
```
I think it's bad to single out 200. We should allow other OK status: https://fetch.spec.whatwg.org/#ok-status (and add tests for them).

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,33 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                          1. Terminate these substeps.
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is `200`, then:
+                          1. [=queue/Enqueue=] |raceNetworkRequestResponse| to |queue|.
+              1. [=If aborted=] and |raceFetchController| is not null, then:
```
"If aborted" can only be used inside "abort when". But this is a sibling step to "abort when". Maybe indentation got messed up?

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,33 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is `200`, then:
+                          1. [=queue/Enqueue=] |raceNetworkRequestResponse| to |queue|.
+              1. [=If aborted=] and |raceFetchController| is not null, then:
+                  1. Let |deserializedError| be the result of [=deserialize a serialized abort reason=] given null and |workerRealm|.
```
null seems like a bad error to use here. I suggest using the fact that [=fetch controller/Abort=] is optional, and do not pass an error.

Also, if this error is web developer visible, we should be sure to test that the correct error object is used. (But I think it is not visible?)

sisidovski replied:
Yes, this is not visible. Updated to call [=fetch controller/Abort=] without passing an error.

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,33 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                  1. Let |deserializedError| be the result of [=deserialize a serialized abort reason=] given null and |workerRealm|.
+                  1. [=fetch controller/Abort=] |raceFetchController| with |deserializedError|.
+              1. Run the following substeps [=in parallel=]:
+                  1. Resolve |preloadResponse| with undefined.
```
You cannot resolve a promise in parallel. I think you can just move this outside of the in parallel block.

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,33 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                  1. Let |fetchHandlerResponse| to the result of [=Create Fetch Event and Dispatch=] with |request|, |registration|, |useHighResPerformanceTimers|, |timingInfo|, |workerRealm|, |reservedClient|, and |preloadResponse|.
+                  1. If |fetchHandlerResponse| is not null and not a [=network error=], and |raceFetchController| is not null, then:
+                      1. Let |deserializedError| be the result of [=deserialize a serialized abort reason=] given null and |workerRealm|.
+                      1. [=fetch controller/Abort=] |raceFetchController| with |deserializedError|.
```
Same comment about how aborting with null is strange and unnecessary.

domenic commented:
Some of the in-parallel stuff you're running into here is a preexisting issue with the spec, it seems. I filed https://github.com/whatwg/fetch/issues/1734. I think my above advice helps avoid making the existing problem worse and so is still worth following.

sisidovski committed:
Fixed passing the error to [=fetch controller/Abort=], indentation, 200 to [=ok status=], updated the caller for fetch |reqeuest|

domenic commented on docs/index.bs:
```diff
@@ -1578,7 +1578,7 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
 
       enum RunningStatus { "running", "not-running" };
-      enum RouterSourceEnum { "cache", "fetch-event", "network" };
+      enum RouterSourceEnum { "cache", "fetch-event", "network", "race-network-and-fetch-handler" };
```
Nit: it may be worth making this multiline now.

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,25 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. Resolve |preloadResponse| with undefined.
+              1. Run the following substeps [=in parallel=]:
+                  1. Let |fetchHandlerResponse| to the result of [=Create Fetch Event and Dispatch=] with |request|, |registration|, |useHighResPerformanceTimers|, |timingInfo|, |workerRealm|, |reservedClient|, and |preloadResponse|.
+                  1. If |fetchHandlerResponse| is not null and not a [=network error=], and |raceFetchController| is not null, [=fetch controller/Abort=] |raceFetchController|.
```
```suggestion
                  1. If |fetchHandlerResponse| is not null and not a [=network error=], and |raceFetchController| is not null, [=fetch controller/abort=] |raceFetchController|.
```

domenic commented on docs/index.bs:
```diff
@@ -3176,6 +3190,29 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
               1. [=fetch controller/Abort=] |preloadFetchController| with |deserializedError|.
       1. Else, resolve |preloadResponse| with undefined.
+      1. Set |response| to the result of [=Create Fetch Event and Dispatch=] with |request|, |registration|, |useHighResPerformanceTimers|, |timingInfo|, |workerRealm|, |reservedClient|, and |preloadResponse|.
+      1. Retrun |response|.
```
Nit: collapse this into a single step, "Return the result of Create Fetch Event and Dispatch with..."

Do the same with "Set _response_ to the result of dequeue queue / Return response".

Then delete "Let response be null" from the start of the algorithm, and change "Output" to just "a response".

domenic commented on docs/index.bs:
```diff
@@ -3176,6 +3190,29 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: |reservedClient|, a [=request/reserved client=].
+      :: |preloadResponse| a [=promise=].
+      : Output
+      :: |response|, a [=/response=]
```
Just declaring the type, not the variable name, seems better to me.

domenic commented on docs/index.bs:
```diff
@@ -3176,6 +3190,29 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: |response|, a [=/response=]
+
+      1. Let |eventCanceled| be false.
```
_response_ is used in this algorithm but never declared.

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,25 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. Run the following substeps [=in parallel=], but [=abort when=] |controller|'s [=fetch controller/state=] is "<code>terminated</code>" or "<code>aborted</code>":
+                  1. Set |raceFetchController| to the result of calling [=fetch=] given |request|, with [=fetch/processResponse=] set to the following steps given a [=/response=] |raceNetworkRequestResponse|:
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is [=ok status=], [=queue/enqueue=] |raceNetworkRequestResponse| to |queue|.
+                  1. [=If aborted=] and |raceFetchController| is not null, [=fetch controller/Abort=] |raceFetchController|.
```
I'm sorry, I gave totally bad advice previously :(. "If aborted" is supposed to be a sibling step, not a child step, of an "abort when" step. I don't know how I missed that.

domenic replied:
```suggestion
                  1. [=If aborted=] and |raceFetchController| is not null, [=fetch controller/abort=] |raceFetchController|.
```

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3150,25 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
-      1. If |request| is a <a>non-subresource request</a>, |request| is a [=navigation request=], |registration|'s [=navigation preload enabled flag=] is set, |request|'s [=request/method=] is \`<code>GET</code>\`, |registration|'s [=active worker=]'s [=set of event types to handle=] [=set/contains=] <code>fetch</code>, and |registration|'s [=active worker=]'s [=all fetch listeners are empty flag ...
+          1. Else if |source| is {{RouterSourceEnum/"race-network-and-fetch-handler"}}, and |request|'s [=request/method=] is \`<code>GET</code>\` then:
+
+              Note: In order to avoid duplicated fetches to |request|, the user agent may reuse |raceNetworkRequestResponse| as the result of {{FetchEvent}}'s [=FetchEvent/potential response=], or the fallback request.
```
This seems like it's specifying a possibly significant and observable step, via a vague non-normative note. Am I understanding correctly?

It seems to me you need to have some place to store the  _raceNetworkRequestResponse_, modify all the call sites that set the potential response, instead.

sisidovski replied:
Yes, your understanding is correct. Let me try writing this behavior in a spec level.

sisidovski replied:
Updated a bit to capture the case when the user agent reuses the network response for the fallback.

The other part is more problematic.

> the user agent may reuse |raceNetworkRequestResponse| as the result of {{FetchEvent}}'s [=FetchEvent/potential response=]

Actually this sentense was wrong. We have a dedupe mechanism for the |request| level, that means if `fetch(e.request)` is called inside the fetch handler, the user agent reuse the response dispatched before. But the response of `fetch(e.request)` is not `[=FetchEvent/potential response=]`. Partial response basically means the returned value in `e.respondWith()`, which may use `fetch(e.request)` response internally, but that's usually differnt data'.

I'm not sure how I can write this behavior in the spec, any advice would be appreciated.

domenic replied:
It sounds like you'd need to add quite a bit of spec machinery here, to parallel the implementation machinery. Probably that would involve a PR to https://fetch.spec.whatwg.org/ to modify the behavior of `fetch()`, or maybe one of the underlying algorithms like main fetch or HTTP-network fetch. And some sort of side table tracking the request -> response that can be used to short circuit the normal fetching, etc.

Basically, how did you implement this in Chromium? You'll need to write up equivalent spec text.

sisidovski replied:
Updated, diff from the last patch is:

1. Added  `request/service-workers race token` and set it when the race network request is created.
2. Added `race response map` in ServiceWorkerGlobalScope. The map takes a token string as a key, and a promise that will be resolved with the response as a value.
3. Added `Lookup Race Response`. That returns the promise if there is an entry in the map, otherwise null. That will be called from `Main fetch` in the fetch spec.

I also updated the [fetch spec](https://github.com/sisidovski/fetch/pull/1).
1. Added `>service-workers race token` as an associated property of request.
2. Added the logic to return `response` early if there is a corresponding entry with the token string.

sisidovski committed:
Collapse some redundant steps, make enum multiline, fixed abot, remove |response|

sisidovski committed:
Add fallback case

sisidovski committed:
Add token and lookup algorithm

sisidovski committed:
Drop incorrect note

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3155,25 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                      1. Set |raceResponse| to [=a new promise=].
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is [=ok status=], then:
+                          1. Resolve |raceResponse| with |raceNetworkRequestResponse|.
+                          1. [=queue/enqueue=] |raceNetworkRequestResponse| to |queue|.
```
```suggestion
                          1. [=queue/Enqueue=] |raceNetworkRequestResponse| to |queue|.
```

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3155,25 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                  1. Set |raceFetchController| to the result of calling [=fetch=] given |request|, with [=fetch/processResponse=] set to the following steps given a [=/response=] |raceNetworkRequestResponse|:
+                      1. Set |raceResponse| to [=a new promise=].
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is [=ok status=], then:
+                          1. Resolve |raceResponse| with |raceNetworkRequestResponse|.
```
```suggestion
                          1. [=Resolve=] |raceResponse| with |raceNetworkRequestResponse|.
```

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3155,25 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is [=ok status=], then:
+                          1. Resolve |raceResponse| with |raceNetworkRequestResponse|.
+                          1. [=queue/enqueue=] |raceNetworkRequestResponse| to |queue|.
+                      1. Otherwise, reject |raceResponse| with a `TypeError`.
```
```suggestion
                      1. Otherwise, [=reject=] |raceResponse| with a `TypeError`.
```

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3155,25 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. [=If aborted=] and |raceFetchController| is not null, [=fetch controller/abort=] |raceFetchController|.
+              1. Resolve |preloadResponse| with undefined.
+              1. Run the following substeps [=in parallel=]:
+                  1. Let |fetchHandlerResponse| to the result of [=Create Fetch Event and Dispatch=] with |request|, |registration|, |useHighResPerformanceTimers|, |timingInfo|, |workerRealm|, |reservedClient|, |preloadResponse|, and |raceResponse|.
```
```suggestion
                  1. Let |fetchHandlerResponse| be the result of [=Create Fetch Event and Dispatch=] with |request|, |registration|, |useHighResPerformanceTimers|, |timingInfo|, |workerRealm|, |reservedClient|, |preloadResponse|, and |raceResponse|.
```

domenic commented on docs/index.bs:
```diff
@@ -3176,6 +3195,30 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: |workerRealm|, a [=relevant realm=] of the [=service worker/global object=].
+      :: |reservedClient|, a [=request/reserved client=].
+      :: |preloadResponse| a [=promise=].
+      :: |raceResponse| a [=promise=].
```
```suggestion
      :: |raceResponse|, a [=promise=]
```

domenic commented on docs/index.bs:
```diff
@@ -3176,6 +3195,30 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: |timingInfo|, a [=service worker timing info=].
+      :: |workerRealm|, a [=relevant realm=] of the [=service worker/global object=].
+      :: |reservedClient|, a [=request/reserved client=].
+      :: |preloadResponse| a [=promise=].
```
```suggestion
      :: |preloadResponse|, a [=promise=]
```

domenic commented on docs/index.bs:
```diff
@@ -3176,6 +3195,30 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: |useHighResPerformanceTimers|, a boolean
+      :: |timingInfo|, a [=service worker timing info=].
+      :: |workerRealm|, a [=relevant realm=] of the [=service worker/global object=].
+      :: |reservedClient|, a [=request/reserved client=].
```
```suggestion
      :: |reservedClient|, a [=request/reserved client=]
```

domenic commented on docs/index.bs:
```diff
@@ -3176,6 +3195,30 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: |registration|, a [=/service worker registration=]
+      :: |useHighResPerformanceTimers|, a boolean
+      :: |timingInfo|, a [=service worker timing info=].
+      :: |workerRealm|, a [=relevant realm=] of the [=service worker/global object=].
```
```suggestion
      :: |workerRealm|, a [=relevant realm=] of the [=service worker/global object=]
```

domenic commented on docs/index.bs:
```diff
@@ -3176,6 +3195,30 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: |request|, a [=/request=]
+      :: |registration|, a [=/service worker registration=]
+      :: |useHighResPerformanceTimers|, a boolean
+      :: |timingInfo|, a [=service worker timing info=].
```
```suggestion
      :: |timingInfo|, a [=service worker timing info=]
```

domenic commented on docs/index.bs:
```diff
@@ -3195,6 +3238,11 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
       1. Else:
           1. Set |workerRealm| to the [=relevant realm=] of the |activeWorker|'s [=service worker/global object=].
           1. Set |eventHandled| to [=a new promise=] in |workerRealm|.
+          1. If |raceResponse| is not null, then:
```
Is it possible raceResponse will be set to a promise in step 13.4.3.1.1 of Handle Fetch, but never resolved with a response by step 13.4.3.1.2.1? Or, it will be rejected because of the fetch abort that occurs in step 13.4.4?

domenic replied:
Now it might be "pending", I think? The map does not say it is allowed to contain "pending", but maybe it should be expanded to allow that?

domenic commented on docs/index.bs:
```diff
@@ -3195,6 +3238,11 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. Let |map| be |activeWorker|'s [=service worker/global object=]'s [=race response map=].
+              1. Let |token| be the result of [=generating a random UUID=].
+              1. [=map/Set=] |map|[|token|] to |raceResponse|.
+              1. Set |request|'s [=request/service-workers race token=] to |token|.
```
Is it possible to just make the map keyed by requests, instead of strings? Then you don't need this token setup.

sisidovski replied:
Is it possible to get the map entry by using a struct (or object) as a key? If so that would be feasible.
On the other hand, the implementation side may become a bit difficult due to the type inconsistency of request.

domenic replied:
Yes. In specs maps can be keyed by anything. There's no need to change the implementation; as long as the spec is observably equivalent, it is OK for them to depart.

But, be sure to distinguish between requests and `Request`s in the spec...

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3155,25 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. Let |raceFetchController| be null.
+              1. Run the following substeps [=in parallel=], but [=abort when=] |controller|'s [=fetch controller/state=] is "<code>terminated</code>" or "<code>aborted</code>":
+                  1. Set |raceFetchController| to the result of calling [=fetch=] given |request|, with [=fetch/processResponse=] set to the following steps given a [=/response=] |raceNetworkRequestResponse|:
+                      1. Set |raceResponse| to [=a new promise=].
```
I'm not sure promises work well here. Promises can only store JavaScript or IDL objects, like `Response`, but you are instead resolving them with a [response](https://fetch.spec.whatwg.org/#concept-response). That isn't allowed.

You could try to wrap the response into a `Response`. I guess that would mean creating a [new](https://webidl.spec.whatwg.org/#new) `Response` whose [response](https://fetch.spec.whatwg.org/#concept-response-response) is set to the _response_. But as explained in a comment later on, I don't think that is the right architecture. In general, I don't think using promises (which are main thread objects, very tied to JavaScript) inside this background-thread processing is a good idea.

sisidovski replied:
Tried to update not to use promise here. Does that look better?

domenic commented on docs/index.bs:
```diff
@@ -3236,6 +3284,10 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
               2. Return a [=network error=].
           1. If |eventHandled| is not null, then [=resolve=] |eventHandled|.
+          1. If |raceResponse| is not null, then:
+              1. [=promise/React=] to |raceResponse|:
```
You can't return from inside the react steps. Also, as discussed above, there is a type mismatch here: promises can only hold JavaScript objects, and you need to return a response, not a `Response`.

sisidovski committed:
Fix incorrect format, update the map key to be request

sisidovski committed:
Do not use promise to manage [=/response=]

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3155,27 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is [=ok status=], then:
+                          1. Set |raceResponse| to |raceNetworkRequestResponse|.
+                          1. [=queue/Enqueue=] |raceNetworkRequestResponse| to |queue|.
+                      1. Otherwise, set |raceResponse| to a `TypeError`.
```
Probably set it to a `[=network error=]`, so that its type stays null, "pending", or response. (network error is a type of response.) Or maybe null?

sisidovski replied:
[=network error=] was my intention. Thanks.

domenic commented on docs/index.bs:
```diff
@@ -3176,6 +3197,30 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: |workerRealm|, a [=relevant realm=] of the [=service worker/global object=]
+      :: |reservedClient|, a [=request/reserved client=]
+      :: |preloadResponse|, a [=promise=]
+      :: |raceResponse|, a null, "<code>pending</code>" or [=/response=]
```
Parameter passing in specs generally follows JavaScript-style conventions. So, you cannot modify a parameter from outside the called algorithm, once it has been passed.

To get around this, create a tiny wrapper struct, with one item. Then the other algorithm can modify the struct item.

sisidovski replied:
> Now it might be "pending", I think? The map does not say it is allowed to contain "pending", but maybe it should be expanded to allow that?

Yes, in L3141, the state is either "pending" or [=/response=]. I'll update the map to accept "pending".

sisidovski replied:
> To get around this, create a tiny wrapper struct, with one item. Then the other algorithm can modify the struct item.

Added a new struct and use it to the parameter, but not fully confident if this direction is correct. Let me know the usage is wrong.

domenic reviewed (commented):
This is looking like it will work :).

domenic commented on docs/index.bs:
```diff
@@ -1079,7 +1079,7 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
     </pre>
 
-    A {{ServiceWorkerGlobalScope}} object represents the global execution context of a [=/service worker=]. A {{ServiceWorkerGlobalScope}} object has an associated <dfn export for="ServiceWorkerGlobalScope">service worker</dfn> (a [=/service worker=]). A {{ServiceWorkerGlobalScope}} object has an associated <dfn for="ServiceWorkerGlobalScope">force bypass cache for import scripts flag</dfn>. It i ...
+    A {{ServiceWorkerGlobalScope}} object represents the global execution context of a [=/service worker=]. A {{ServiceWorkerGlobalScope}} object has an associated <dfn export for="ServiceWorkerGlobalScope">service worker</dfn> (a [=/service worker=]). A {{ServiceWorkerGlobalScope}} object has an associated <dfn for="ServiceWorkerGlobalScope">force bypass cache for import scripts flag</dfn>. A {{ ...
```
I think the map values are now "response"s?

sisidovski replied:
"response"s or "pending"? I updated it with the new struct though.

sisidovski committed:
Added the race response struct

sisidovski commented on docs/index.bs:
```diff
@@ -3873,6 +3926,27 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Input
+      :: |request|, a [=request=]
+      : Output
+      :: a [=race response=].
```
As this is called from the fetch spec, if possible, perhaps it's convenient to set just `"<code>pending</code>", [=/response=], or null` as an output.

domenic replied:
Yes, I think unwrapping the struct would be better, so that the Fetch spec doesn't need to know about it. (And then you can avoid `export`ing it.)

Also, is Fetch really going to treat all three outcomes differently? Or maybe if it's still pending at the time of the call, it will treat pending the same as null? If so the you could handle that in this algorithm too.

sisidovski replied:
Thanks! Updated to return [=/response=] or null. Fetch is going to treat if there is an map entry, which returns [=/response=], or no entry.

domenic commented on docs/index.bs:
```diff
@@ -1079,7 +1079,9 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
-    A {{ServiceWorkerGlobalScope}} object represents the global execution context of a [=/service worker=]. A {{ServiceWorkerGlobalScope}} object has an associated <dfn export for="ServiceWorkerGlobalScope">service worker</dfn> (a [=/service worker=]). A {{ServiceWorkerGlobalScope}} object has an associated <dfn for="ServiceWorkerGlobalScope">force bypass cache for import scripts flag</dfn>. It i ...
+    A {{ServiceWorkerGlobalScope}} object represents the global execution context of a [=/service worker=]. A {{ServiceWorkerGlobalScope}} object has an associated <dfn export for="ServiceWorkerGlobalScope">service worker</dfn> (a [=/service worker=]). A {{ServiceWorkerGlobalScope}} object has an associated <dfn for="ServiceWorkerGlobalScope">force bypass cache for import scripts flag</dfn>. A {{ ...
+
+    A <dfn export id="dfn-race-response">race response</dfn> is a [=struct=] used to contain the network response when {{RouterSourceEnum/"race-network-and-fetch-handler"}} performs. It has [=/response=], or "<code>pending</code>" as an [=struct/item=].
```
You should give the struct item a name. E.g. ``It has a <dfn for="race response">value</dfn>, which is a [=/response=] or "`pending`".``

I think the value can also be null, though?

sisidovski replied:
Yes. That can take null. And thank for putting samples. Updated.

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3157,27 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+          1. Else if |source| is {{RouterSourceEnum/"race-network-and-fetch-handler"}}, and |request|'s [=request/method=] is \`<code>GET</code>\` then:
+              1. Let |queue| be an empty [=queue=] of [=/response=].
+              1. Let |raceFetchController| be null.
+              1. Set |raceResponse| to "<code>pending</code>".
```
Here it would be: ``Set |raceResponse|'s [=race response/value=] to "`pending`".``

domenic commented on docs/index.bs:
```diff
@@ -3154,7 +3157,27 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. Run the following substeps [=in parallel=], but [=abort when=] |controller|'s [=fetch controller/state=] is "<code>terminated</code>" or "<code>aborted</code>":
+                  1. Set |raceFetchController| to the result of calling [=fetch=] given |request|, with [=fetch/processResponse=] set to the following steps given a [=/response=] |raceNetworkRequestResponse|:
+                      1. If |raceNetworkRequestResponse|'s [=response/status=] is [=ok status=], then:
+                          1. Set |raceResponse| to |raceNetworkRequestResponse|.
```
Here it would be: `Set |raceResponse|'s [=race response/value=] to |raceNetworkRequestResponse|`.

domenic commented on docs/index.bs:
```diff
@@ -3106,19 +3113,15 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
       1. Let |workerRealm| be null.
-      1. Let |eventHandled| be null.
       1. Let |timingInfo| be a new [=service worker timing info=].
+      1. Let |raceResponse| be null.
```
Here it would be: ``Set |raceResponse| to a [=race response=] whose [=race response/value=] is null.``

sisidovski replied:
I believe this is `Let`?

domenic reviewed (commented):
I gave a few samples of how to do the struct, but I did not comment on every access.

sisidovski committed:
Update race response struct

sisidovski committed:
Remove unconsumed map entry

sisidovski commented on docs/index.bs:
```diff
@@ -3231,11 +3280,15 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
 
       1. Wait for |task| to have executed or for |handleFetchFailed| to be true.
       1. If |shouldSoftUpdate| is true, then [=in parallel=] run the [=Soft Update=] algorithm with |registration|.
+      1. If |raceResponseMap|[|request|] [=map/exists=], [=map/remove=] |raceResponseMap|[|request|].
```
This step is needed to remove the map entry, which is not consumed by Fetch. The fetch handler may not have `fetch(e.request)` e.g. fallback, offline etc. Otherwise map entries are infinitely added.

mkruisselbrink commented on docs/index.bs:
```diff
@@ -3154,7 +3157,28 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. Let |queue| be an empty [=queue=] of [=/response=].
+              1. Let |raceFetchController| be null.
+              1. Set |raceResponse|'s [=race response/value=] to "<code>pending</code>".
+              1. [=map/Set=] |activeWorker|'s [=service worker/global object=]'s [=race response map=][|request|] to |raceResponse|.
```
It might be simpler if the race response map is associates with activeWorker (i.e. a service worker) rather than its global scope? Especially as at this point in the algorithm the worker hasn't been started yet, so global object might very well be null.

sisidovski replied:
That's a good point. How about keep the global scope having the map, and move this [=map/set=] to the step 16.2 in _[Create Fetch Event and Dispatch](https://pr-preview.s3.amazonaws.com/sisidovski/ServiceWorkerForStaticRoutingAPI/pull/10.html#create-fetch-event-and-dispatch-algorithm)_? At this point, _[Run Service Worker](https://pr-preview.s3.amazonaws.com/sisidovski/ServiceWorkerForStaticRoutingAPI/pull/10.html#run-service-worker)_ was already called to we can expect the global object is ready.

sisidovski replied:
Updated as above. Let me know if this doesn't solve your concern.

sisidovski committed:
Move adidng an entry after `run service worker`

sisidovski committed:
Merge branch 'static_routing_api' into patch-2

domenic commented on docs/index.bs:
```diff
@@ -1082,7 +1082,9 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
     </pre>
 
-    A {{ServiceWorkerGlobalScope}} object represents the global execution context of a [=/service worker=]. A {{ServiceWorkerGlobalScope}} object has an associated <dfn export for="ServiceWorkerGlobalScope">service worker</dfn> (a [=/service worker=]). A {{ServiceWorkerGlobalScope}} object has an associated <dfn for="ServiceWorkerGlobalScope">force bypass cache for import scripts flag</dfn>. It i ...
+    A {{ServiceWorkerGlobalScope}} object represents the global execution context of a [=/service worker=]. A {{ServiceWorkerGlobalScope}} object has an associated <dfn export for="ServiceWorkerGlobalScope">service worker</dfn> (a [=/service worker=]). A {{ServiceWorkerGlobalScope}} object has an associated <dfn for="ServiceWorkerGlobalScope">force bypass cache for import scripts flag</dfn>. A {{ ...
```
```suggestion
    A {{ServiceWorkerGlobalScope}} object represents the global execution context of a [=/service worker=]. A {{ServiceWorkerGlobalScope}} object has an associated <dfn export for="ServiceWorkerGlobalScope">service worker</dfn> (a [=/service worker=]). A {{ServiceWorkerGlobalScope}} object has an associated <dfn for="ServiceWorkerGlobalScope">force bypass cache for import scripts flag</dfn>. A {{ServiceWorkerGlobalScope}} object has an associated <dfn for="ServiceWorkerGlobalScope">race response map</dfn> which is an [=ordered map=] where the [=map/keys=] are [=/requests=] and the [=map/values=] are [=race response=]. It is initially unset.
```

I think this will fix the linking...

domenic commented on docs/index.bs:
```diff
@@ -3202,6 +3205,26 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                          1. [=queue/Enqueue=] |raceNetworkRequestResponse| to |queue|.
+                      1. Otherwise, set |raceResponse|'s [=race response/value=] to a [=network error=].
+              1. [=If aborted=] and |raceFetchController| is not null, then:
+                  1. [=fetch controller/abort=] |raceFetchController|.
```
```suggestion
                  1. [=fetch controller/Abort=] |raceFetchController|.
```

domenic commented on docs/index.bs:
```diff
@@ -3277,11 +3329,15 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
               2. Return a [=network error=].
           1. If |eventHandled| is not null, then [=resolve=] |eventHandled|.
+          1. If |raceResponse|'s [=race response/value=] is not null, then:
+              1. Wait until |raceResponse|'s [=race response/value=] is not "<code>pending</code>"
```
```suggestion
              1. Wait until |raceResponse|'s [=race response/value=] is not "<code>pending</code>".
```

domenic commented on docs/index.bs:
```diff
@@ -3277,11 +3329,15 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
           1. If |eventHandled| is not null, then [=resolve=] |eventHandled|.
+          1. If |raceResponse|'s [=race response/value=] is not null, then:
+              1. Wait until |raceResponse|'s [=race response/value=] is not "<code>pending</code>"
+              1. If |raceResponse|'s [=race response/value=] is [=/response=], return |raceResponse|'s [=race response/value=].
```
```suggestion
              1. If |raceResponse|'s [=race response/value=] is a [=/response=], return |raceResponse|'s [=race response/value=].
```

domenic commented on docs/index.bs:
```diff
@@ -3922,6 +3978,28 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Input
+      :: |request|, a [=/request=]
+      : Output
+      :: a [=/response=], or null.
```
```suggestion
      :: a [=/response=] or null
```

domenic reviewed (approved):
LGTM!

sisidovski committed:
Fix editorial issues
